### PR TITLE
Create affiliate stats table and activity insert trigger

### PR DESCRIPTION
**Pull request by youngkidwarrior**

### CODE REVIEW

**Comment by youngkidwarrior** [comment]
* **#717** <a href="https://app.graphite.dev/github/pr/0xsend/sendapp/717?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `dev`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @youngkidwarrior and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: supabase/migrations/20240919082056_create_affiliate_stats_table.sql [hunk 1]
```diff
@@ -0,0 +1,44 @@
+CREATE TABLE "public"."affiliate_stats"(
+    "paymaster_tx_count" numeric NOT NULL DEFAULT '0' ::numeric,
```
Comment (0xBigBoss) [inline supabase/migrations/20240919082056_create_affiliate_stats_table.sql:2]: add created_at and updated_at pls

File: supabase/migrations/20240919082056_create_affiliate_stats_table.sql [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+ALTER TABLE "public"."affiliate_stats"
+    ADD CONSTRAINT "affiliate_stats_pkey" PRIMARY KEY USING INDEX "affiliate_stats_pkey";
+
+ALTER TABLE "public"."affiliate_stats"
+    ADD CONSTRAINT "affiliate_stats_user_id_fkey" FOREIGN KEY (user_id) REFERENCES profiles(id) ON UPDATE CASCADE ON DELETE CASCADE NOT valid;
+
+ALTER TABLE "public"."affiliate_stats" validate CONSTRAINT "affiliate_stats_user_id_fkey";
+
+GRANT DELETE ON TABLE "public"."affiliate_stats" TO "service_role";
```
Comment (0xBigBoss) [inline supabase/migrations/20240919082056_create_affiliate_stats_table.sql:19]: remove these grants/revokes. no need.

File: supabase/migrations/20240919082056_create_affiliate_stats_table.sql [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+GRANT TRIGGER ON TABLE "public"."affiliate_stats" TO "service_role";
+
+GRANT TRUNCATE ON TABLE "public"."affiliate_stats" TO "service_role";
+
+GRANT UPDATE ON TABLE "public"."affiliate_stats" TO "service_role";
+
+INSERT INTO affiliate_stats(user_id, paymaster_tx_count)
+SELECT
+    au.id AS user_id,
+    COALESCE(COUNT(activity.*) FILTER (WHERE activity.from_user_id IS NOT NULL
```
Comment (0xBigBoss) [inline supabase/migrations/20240919082056_create_affiliate_stats_table.sql:36]: this isn't filtering transfers to the paymaster btw. Why not use the `send_account_transfers` table? I don't have the mainnet address handy, but we will most likely want to create a `private.paymasters` table that has our paymaster addresses in it + their chain ids.

Then, we can join or do a where `t in (select address from private.paymaster)`.

File: supabase/migrations/20240919092852_update_affiliate_stats_on_insert_trigger.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
+SET check_function_bodies = OFF;
+
+CREATE OR REPLACE FUNCTION public.update_affiliate_stats_on_activity_insert()
+    RETURNS TRIGGER
+    LANGUAGE plpgsql
+    AS $function$
+BEGIN
+    -- Check if the new activity matches the conditions for incrementing paymaster_tx_count
+    IF NEW.from_user_id IS NOT NULL AND NEW.to_user_id IS NULL AND NEW.event_name = 'send_account_transfers' THEN
```
Comment (0xBigBoss) [inline supabase/migrations/20240919092852_update_affiliate_stats_on_insert_trigger.sql:9]: where are we checking the referrer user id?

File: supabase/migrations/20240919082056_create_affiliate_stats_table.sql (hunk 3 above)
Comment (youngkidwarrior) [reply supabase/migrations/20240919082056_create_affiliate_stats_table.sql:36]: The activity table is a bounty of information for affiliates. We are going to pull more stats and we can do them all from activity. It's extremely easy to query and run triggers on

Comment (0xBigBoss) [reply supabase/migrations/20240919082056_create_affiliate_stats_table.sql:36]: withdrawals. needs to filter `t = <paymaster address>`

**Comment by youngkidwarrior** [comment]
### Merge activity

* **Sep 22, 4:03 AM EDT**: @youngkidwarrior started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/0xsend/sendapp/717).

**Pull request merged**